File: src/main/java/de/officeryoda/fhysics/engine/FhysicsCore.java
package de.officeryoda.fhysics.engine;

import de.officeryoda.fhysics.objects.Circle;
import de.officeryoda.fhysics.rendering.FhysicsObjectDrawer;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

public class FhysicsCore {

    public static final Border BORDER = new Border(0, 600, 0, 400);
    private static int objectCount;
    @Getter
    private final List<Circle> fhysicsObjects;
    private final Vector2 gravity = new Vector2(0, -9.81);
    private final int updatesPerSecond = 500;
    private final double updatesIntervalSeconds = 1.0 / updatesPerSecond;
    @Setter
    private FhysicsObjectDrawer drawer;

    public FhysicsCore() {
        this.fhysicsObjects = new ArrayList<>();

        fhysicsObjects.add(new Circle(new Vector2(50, 200), 20));
        fhysicsObjects.add(new Circle(new Vector2(400, 190), 20));

        fhysicsObjects.add(new Circle(new Vector2(50, 305), 5));
        fhysicsObjects.add(new Circle(new Vector2(250, 150), 12));
        fhysicsObjects.add(new Circle(new Vector2(100, 200), 15));
        fhysicsObjects.add(new Circle(new Vector2(500, 360), 10));
//        fhysicsObjects.add(new Box(new Vector2(300, 400), 50, 100));
//        fhysicsObjects.add(new Box(new Vector2(300, 31), 50, 100));

        fhysicsObjects.getFirst().setVelocity(new Vector2(200, 0));
    }

    public static int nextId() {
        return objectCount++;
    }

    public void startUpdateLoop() {
        Timer updateTimer = new Timer(true);
        // time in ms

        int updateIntervalMillis = (int) (1f / updatesPerSecond * 1000);
        updateTimer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                update();
            }
        }, 0, updateIntervalMillis);
    }

    private void update() {
        fhysicsObjects.forEach(obj -> {
//            obj.applyGravity(updatesIntervalSeconds, gravity);
            obj.applyGravity(updatesIntervalSeconds, Vector2.zero());
            checkBorderCollision(obj);
            checkObjectCollision(obj);
        });
        drawer.repaintObjects();
    }

    private void checkObjectCollision(Circle obj1) {
        fhysicsObjects.forEach(obj2 -> {
            if(obj1.getId() == obj2.getId()) return;
            CollisionHandler.handleElasticCollision(obj1, obj2);
        });
    }

    private void checkBorderCollision(Circle circle) {
        Vector2 pos = circle.getPosition();
        Vector2 velocity = circle.getVelocity();
        double radius = circle.getRadius();

        // check top/bottom border
        if(pos.getY() + radius > BORDER.topBorder()) {
            velocity.setY(-velocity.getY());
//            velocity.set(Vector2.zero());
            pos.setY(BORDER.topBorder() - radius);
//            velocity.setX(Math.random() * 400 - 200);
        } else if(pos.getY() - radius < BORDER.bottomBorder()) {
            velocity.setY(-velocity.getY());
            pos.setY(BORDER.bottomBorder() + radius);
//            velocity.setX(Math.random() * 400 - 200);
        }

        // check left/right border
        if(pos.getX() - radius < BORDER.leftBorder()) {
            velocity.setX(-velocity.getX());
            pos.setX(BORDER.leftBorder() + radius);
        } else if(pos.getX() + radius > BORDER.rightBorder()) {
            velocity.setX(-velocity.getX());
            pos.setX(BORDER.rightBorder() - radius);
        }
    }
}

File: src/main/java/de/officeryoda/fhysics/engine/FhysicsCore.kt
package de.officeryoda.fhysics.engine

import de.officeryoda.fhysics.engine.Settings.SUB_STEPS
import de.officeryoda.fhysics.engine.Settings.UPDATES_PER_SECOND
import de.officeryoda.fhysics.engine.datastructures.BoundingBox
import de.officeryoda.fhysics.engine.datastructures.QuadTree
import de.officeryoda.fhysics.engine.math.Vector2
import de.officeryoda.fhysics.engine.objects.FhysicsObject
import de.officeryoda.fhysics.engine.util.Stopwatch
import de.officeryoda.fhysics.engine.util.times
import de.officeryoda.fhysics.visual.Renderer
import de.officeryoda.fhysics.visual.SceneListener
import java.util.*
import java.util.concurrent.locks.ReentrantLock
import kotlin.math.max

/**
 * The core of the Fhysics engine.
 * This class contains all the important variables and functions for the simulation.
 */
object FhysicsCore {

    /** A Thread lock to prevent simulating and rendering at the same time */
    val RENDER_LOCK = ReentrantLock()

    /** The Bounds of the simulation */
    var border: BoundingBox = BoundingBox(0f, 0f, 100f, 100f) // x and y must be 0.0

    /** The time step used for the simulation */
    var dt: Float = 1.0f / (UPDATES_PER_SECOND * SUB_STEPS) * Settings.timeScale

    /** The amount of updates that have been performed since the start of the simulation */
    var updateCount = 0 // Includes all sub steps

    /** Whether the simulation is running */
    var running: Boolean = true

    /** A stopwatch to measure the time per update */
    val updateStopwatch = Stopwatch()

    fun startEverything() {
        // Start the rendering thread
        Thread { Renderer().launch() }.start()
        // Start the simulation
        startUpdateLoop()
    }

    private fun startUpdateLoop() {
        val updateIntervalMillis: Long = (1f / UPDATES_PER_SECOND * 1000).toLong()

        Timer("Fhysics-Core", true).scheduleAtFixedRate(object : TimerTask() {
            override fun run() {
                if (running) {
                    // Don't render while updating to prevent concurrentModificationExceptions
                    RENDER_LOCK.lock()
                    update()
                    RENDER_LOCK.unlock()
                }
            }
        }, 0, updateIntervalMillis)
    }

    /**
     * Updates the simulation by one step.
     */
    fun update() {
        updateStopwatch.start()

        // Load a new scene if requested
        SceneManager.loadPendingScene()

        // Rebuild must happen before processing pending operations
        QuadTree.rebuild()
        QuadTree.processPendingOperations()

        // Update objects multiple times per update
        repeat(SUB_STEPS) {
            QuadTree.update()

            SceneListener.pullObject()

            updateCount++
        }

        updateStopwatch.stop()
    }

    fun spawn(vararg objects: FhysicsObject): List<FhysicsObject> {
        return spawn(objects.toList())
    }

    private fun spawn(objects: List<FhysicsObject>): List<FhysicsObject> {
        for (obj: FhysicsObject in objects) {
            QuadTree.insert(obj)
        }

        return objects
    }

    /**
     * Calculates the gravity at the given [position][pos].
     */
    fun gravityAt(pos: Vector2): Vector2 {
        if (Settings.gravityType == GravityType.DIRECTIONAL) {
            return Settings.gravityDirection
        } else {
            val direction: Vector2 = Settings.gravityPoint - pos
            val sqrDistance: Float =
                max(1f, direction.sqrMagnitude()) // Prevent high forces when the object is close to the gravity point
            return Settings.gravityPointStrength / sqrDistance * direction.normalized()
        }
    }
}
